### PR TITLE
Remove duplicate `*cat` implementations

### PR DESCRIPTION
**Pull request by lgoettgens**
Resolves https://github.com/Nemocas/AbstractAlgebra.jl/issues/1466.

The `*cat(::MatElem...)` are bad duplicates to methods in `src/Matrix.jl`.
The `*cat(::Vector{T}) where T < MatElem` shouldn't have been there in the first place. They originate form Nemo/Hecke. They shadow the `*cat(::Vector...)` dispatch from base with a 1-elem vararg, and something similar isn't available anywhere else either. So I suggest to remove them. Let's talk about it this afternoon.

### CODE REVIEW

**Comment by thofma** [comment]
If we remove them, we need an equally fast alternative for concatenating a list of matrices.

**Comment by lgoettgens** [comment]
I think you don't like `vcat(vec...)` due to compilation overhead, right?

But there is already `_vcat` and `_hcat` that each take a Vector of matrices. They are part of the internals of our `*cat` implementation in `src/Matrix.jl`. I think these are exactly what you want. We can just export them or find a better name.

**Comment by thofma** [comment]
This is exactly the point. No should ever splat a vector. Base says to use `reduce(vcat, Vector{<:Matrix})`, because there is a special method for that. Ideally we have the same, as we always try to do. It is also advertised in the `vcat` docstring.

**Comment by lgoettgens** [comment]
> This is exactly the point. One should never splat a vector. Base says to use `reduce(vcat, Vector{<:Matrix})`, because there is a special method for that. Ideally we have the same, as we always try to do. It is also advertised in the `vcat` docstring.

`_vcat` and `_hcat` are used for a reduce overload as well.

**Comment by thofma** [comment]
Yes, they should be removed too:
```
julia> reduce(vcat, Union{}[])
ERROR: MethodError: reduce(::typeof(vcat), ::Vector{Union{}}) is ambiguous.

Candidates:
  reduce(::typeof(vcat), A::AbstractVector{<:AbstractVecOrMat})
    @ Base abstractarray.jl:1705
  reduce(::typeof(vcat), A::AbstractVector{<:MatrixElem})
    @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/7siWW/src/Matrix.jl:6499

Possible fix, define
  reduce(::typeof(vcat), ::AbstractVector{Union{}})

Stacktrace:
 [1] top-level scope
   @ REPL[90]:1
```

**Comment by thofma** [comment]
Maybe to summarise my point. All these methods need to go. But I don't want to do it unless we have a good substitute, which is not a new function with some obscure name differing from Base.

File: src/NemoStuff.jl [hunk 1]
```diff
@@ -176,129 +176,6 @@ function sub(M::Generic.Mat, rows::AbstractUnitRange{Int}, cols::AbstractUnitRan
...
-            for k = 1:nrows(i)
-                M[k, s+j] = i[k, j]
-            end
-        end
-        s += ncols(i)
-    end
-    return M
-end
-
-function Base.hcat(A::MatElem...)
```
Comment (fingolfin) [inline src/NemoStuff.jl:225]: Do we really need to remove this? I just tried with removing just the preceding methods, and `vcat(Union{}[])` as well as `vcat()` worked just fine.

Same question for the other methods taking `::MatElem...` args

Comment (lgoettgens) [reply src/NemoStuff.jl:225]: They already exist in better in `src/Matrix.jl`. So these removed methods here get never actually called

**Comment by thofma** [comment]
We plan to go forward with this PR here, but I think we decided to first fix all the call sites to make the transition a bit smoother.

**Comment by lgoettgens** [comment]
https://github.com/thofma/Hecke.jl/pull/1253 will fix many obvious call sites in Hecke.

**Comment by fingolfin** [comment]
Is this ready for merging into a breaking AA release? Or anything that should be done before (then perhaps convert it to draft)?

**Comment by lgoettgens** [comment]
From my POV this can be merged. @thofma do you have any objections?

**Pull request merged**